api/client/client.go

```diff
@@ -0,0 +1,109 @@
+package client
+
+import (
+	"context"
+	"io"
+
+	"github.com/andres-teleport/overseer/api"
+	"github.com/andres-teleport/overseer/api/authentication"
+	"google.golang.org/grpc"
+)
+
+type Client struct {
+	client api.JobworkerServiceClient
+}
+
+type Status struct {
+	Status   string
+	ExitCode int64
+}
+
+func NewClient(serverAddr, keyFile, certFile, caFile string) (*Client, error) {
+	creds, err := authentication.NewClientTransportCredentials(keyFile, certFile, caFile)
+	if err != nil {
+		return nil, err
+	}
+
+	conn, err := grpc.Dial(serverAddr, grpc.WithTransportCredentials(creds))
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{client: api.NewJobworkerServiceClient(conn)}, nil
+}
+
+func (c *Client) Start(command string, arguments ...string) (string, error) {
+	job := &api.Job{
+		Command:   command,
+		Arguments: arguments,
+	}
+
+	jobID, err := c.client.Start(context.Background(), job)
+	if err != nil {
+		return "", err
+	}
+
+	return jobID.Id, nil
+}
+
+func (c *Client) Stop(jobID string) error {
+	_, err := c.client.Stop(context.Background(), &api.JobID{Id: jobID})
+	return err
+}
+
+func (c *Client) Status(jobID string) (Status, error) {
+	statusResp, err := c.client.Status(context.Background(), &api.JobID{Id: jobID})
+	if err != nil {
+		return Status{}, err
+	}
+
+	return Status{
+		Status:   statusResp.Status.String(),
+		ExitCode: statusResp.ExitCode,
+	}, nil
+}
+
+func copyStream(streamer api.JobworkerService_StdOutClient, w *io.PipeWriter) {
+	for eof := false; !eof; {
+		chunk, err := streamer.Recv()
+		if err == io.EOF {
+			eof = true
+		} else if err != nil {
+			w.CloseWithError(err)
+		}
+
+		if chunk != nil {
+			if _, err := w.Write(chunk.Output); err != nil {
+				w.CloseWithError(err)
+				return
+			}
+		}
+	}
+
+	w.Close()
+}
+
+func (c *Client) StdOut(jobID string) (*io.PipeReader, error) {
+	client, err := c.client.StdOut(context.Background(), &api.JobID{Id: jobID})
+	if err != nil {
+		return nil, err
+	}
+
+	pr, pw := io.Pipe()
+	go copyStream(client, pw)
+
+	return pr, nil
+}
```

Comment:
How do you handle situation when user of the library want to stop ongoing log stream ?

Reply:
Perhaps I'm not understanding correctly but the reader could just close the returned `*io.PipeReader` if no more data needs to be read, this will make `w.Write(chunk.Output)` fail in `copyStream` and so it will close the writing end of the pipe.

Reply:
Wouldn't the goroutine leak if the client is stuck on `streamer.Recv()` though?

Reply:
It won't really leak, but it will be stay around for longer than strictly necessary, that is, until the job finishes.